### PR TITLE
[js/web/training] added end-to-end tests

### PR DESCRIPTION
**Pull request by carzh**
## Summary
* following inference's [set-up for end-to-end tests](https://github.com/microsoft/onnxruntime/tree/main/js/web/test/e2e), created an end-to-end test runner for training
* this test runner copies testdata from the [trainingapi folder](https://github.com/microsoft/onnxruntime/tree/main/onnxruntime/test/testdata/training_api) 
* then runs two tests (training session with evalModel & optimizer model, and training session with the minimum options), and tests if the ORT-web training package encompasses inference
  * these tests check 
    * createTrainingSession
    * runTrainStep
    * runOptimizerStep if applicable
    * the parameters methods (getParametersSize, loadParametersBuffer, and getContiguousParameters)

## TL;DR
* [`js/web/test/training/e2e/run.js`](https://github.com/microsoft/onnxruntime/compare/main...carzh:onnxruntime:carzh/training-e2e-runner?expand=1#diff-c1359c4d401f9ba69e937814219cefe5fd11b151a6ffd084c641af3c82e8216c) is responsible for setting up and running the end to end tests
*  [`js/web/test/training/e2e/common.js`](https://github.com/microsoft/onnxruntime/compare/main...carzh:onnxruntime:carzh/training-e2e-runner?expand=1#diff-ee5452491b7b2563d175d13d81d10f2323b12b18589aa4c5798962a8b904a4a8) contains the test function definitions (`testInferenceFunction`, `testTrainingFunctionMin`, `testTrainingFunctionAll`)

## Flow
* entrypoint: user runs the following command in the terminal: `npm run test:training:e2e`
* [`js/web/package.json`](https://github.com/microsoft/onnxruntime/compare/main...carzh:onnxruntime:carzh/training-e2e-runner?expand=1#diff-79275844e75c3c410bb3a71c7f59b2b633e5a3e975c804ffc47220025084da28) was modified to include an npm script that will run `run.js` which will run the end to end tests
* [`js/web/test/training/e2e/run.js`](https://github.com/microsoft/onnxruntime/compare/main...carzh:onnxruntime:carzh/training-e2e-runner?expand=1#diff-c1359c4d401f9ba69e937814219cefe5fd11b151a6ffd084c641af3c82e8216c) is responsible for
  * detecting and installing local tarball packages of ORT-web
  * copying training data to the `js/web/training/e2e/data` folder
  * starting two Karma processes. Karma is a test runner framework that simulates testing in the browser. 
    * In this case, the tests happen in Chrome. We can configure the tests to run in Edge and other browsers in the future. 
    * one of these karma processes is self-hosted, meaning it pulls the ORT-web package from local
    * the other karma process is not self-hosted, meaning it pulls the ORT-web package from another source. In this case, we start an http server that serves the ORT-web binaries.
      * [`js/web/test/training/e2e/simple-http-server.js`](https://github.com/microsoft/onnxruntime/compare/main...carzh:onnxruntime:carzh/training-e2e-runner?expand=1#diff-f798ab485f3ec26c299fe5b2923574c9e4b090200ba20d490bbf6c183286993c) is responsible for starting the HTTP server and serving the ORT binary files. This code almost identical to the same code in the inference E2E tests.
* [`js/web/test/training/e2e/karma.conf.js`](https://github.com/microsoft/onnxruntime/compare/main...carzh:onnxruntime:carzh/training-e2e-runner?expand=1#diff-436cfe8f670c768a04895bd4a1874a5e033f85e0e2d84941c62ff1f7c30a9f28) Karma configuration file that specifies what happens when a karma process is started. The config specifies Mocha as the testing framework, which will go through all the loaded files and run any tests that exist
* [`js/web/test/training/e2e/browser-test-wasm.js`](https://github.com/microsoft/onnxruntime/compare/main...carzh:onnxruntime:carzh/training-e2e-runner?expand=1#diff-13b6155e106dddc7b531ef671186e69b2aadb8a0f4b2f3001db0991567d78221) File that contains the tests that Mocha will pick up on and run. 
* The test functions (such as testInference and testTrainingFunctionAll) are defined in [`js/web/test/training/e2e/common.js`](https://github.com/microsoft/onnxruntime/compare/main...carzh:onnxruntime:carzh/training-e2e-runner?expand=1#diff-ee5452491b7b2563d175d13d81d10f2323b12b18589aa4c5798962a8b904a4a8).

## Notes
* I followed the [tests for training core](https://github.com/microsoft/onnxruntime/blob/b023de0bfc7acb2404dfdcc4adc060b7b72fdaa1/orttraining/orttraining/test/training_api/core/training_api_tests.cc) where they randomly generated input for the training session
* E2E tests are triggered by running `npm run test:training:e2e` -- suggestions for alternative script names are appreciated!!!

## Motivation and Context
- adding training bindings for web

### CODE REVIEW

File: js/web/test/training/e2e/karma.conf.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (askhade) [inline js/web/test/training/e2e/karma.conf.js:1]: @fs-eire can you review this file?

Comment (fs-eire) [reply js/web/test/training/e2e/karma.conf.js:1]: sure. Will take a look today.

File: js/web/test/training/e2e/run.js [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+function prepareTrainingDataByCopying() {
+  const dirs = fs.readdirSync(TRAINING_DATA_FOLDER);
+
+  for (let i = 0; i < dirs.length; i++) {
+    const d = dirs[i];
+    if (fs.statSync(path.join(TRAINING_DATA_FOLDER, d)).isFile()) {
+      const src = path.join(TRAINING_DATA_FOLDER, d);
+      const dest = path.join(TRAININGDATA_DEST, d);
+      fs.copyFile(src, dest);
```
Comment (fs-eire) [inline js/web/test/training/e2e/run.js:146]: use sync copy. otherwise it's possible that when the test is running, the file is not yet ready.

File: js/web/test/training/e2e/karma.conf.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+      {pattern: TEST_MAIN},
+      {pattern: './node_modules/onnxruntime-web/dist/*.wasm', included: false, nocache: true},
+      {pattern: './data/*', included: false},
+    ],
+    plugins: [require('@chiragrupani/karma-chromium-edge-launcher'), ...config.plugins],
+    proxies: {
+      '/model.onnx': '/base/model.onnx',
+      '/data/': '/base/data/',
+      '/test-wasm-path-override/ort-wasm.wasm': '/base/node_modules/onnxruntime-web/dist/ort-wasm.wasm',
+      '/test-wasm-path-override/renamed.wasm': '/base/node_modules/onnxruntime-web/dist/ort-wasm.wasm',
```
Comment (fs-eire) [inline js/web/test/training/e2e/karma.conf.js:34]: should be using ort-training-wasm-simd.wasm instead of ort-wasm.wasm

this also applies to other places in this PR.

File: js/web/test/training/e2e/run.js [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+async function delay(ms) {
+  return new Promise(function(resolve) {
+    setTimeout(function() {
+      resolve();
+    }, ms);
+  });
+}
+
+function prepareTrainingDataByCopying() {
```
Comment (fs-eire) [inline js/web/test/training/e2e/run.js:138]: use fs.copySync() will make these code much simpler. (https://github.com/jprichardson/node-fs-extra/blob/HEAD/docs/copy-sync.md)

File: js/web/test/training/e2e/run.js [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+const startServer = require('./simple-http-server');
+const minimist = require('minimist');
+
+// copy whole folder to out-side of <ORT_ROOT>/js/ because we need to test in a folder that no `package.json` file
+// exists in its parent folder.
+// here we use <ORT_ROOT>/build/js/e2e/ for the test
+
+const TEST_E2E_SRC_FOLDER = __dirname;
+const JS_ROOT_FOLDER = path.resolve(__dirname, '../../../..');
+const TEST_E2E_RUN_FOLDER = path.resolve(JS_ROOT_FOLDER, '../build/js/e2e');
```
Comment (fs-eire) [inline js/web/test/training/e2e/run.js:18]: suggest to use a different folder to the inference E2E test. for example, '../build/js/e2e_training'. Don't use a subfolder of it like '../build/js/e2e/training'.

File: js/web/test/training/e2e/common.js [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+  assertStrictEquals(originalParams.data[0], -0.025190064683556557);
+  assertStrictEquals(originalParams.data[2000], -0.034044936299324036);
+
+  const results = await runTrainStepAndCheck(trainingSession, feeds);
+
+  await trainingSession.runOptimizerStep(feeds);
+  feeds = {"input-0": input0, "labels": labels};
+  // check getContiguousParameters after optimizerStep -- that the parameters have been updated
+  const optimizedParams = await trainingSession.getContiguousParameters();
+  assert(originalParams.data != optimizedParams.data);
```
Comment (fs-eire) [inline js/web/test/training/e2e/common.js:222]: This assert may not do what you want. If you check `obj1 != obj2`, it basically checks if they are the same object. Even if their inside value are same, this expression is always true if they are not the same object. You may want to check things like `originalParams.data[0] !== optimizedParams.data[0]` or so.

File: js/web/test/training/e2e/karma.conf.js (hunk 3 above)
Comment (carzh) [reply js/web/test/training/e2e/karma.conf.js]: removed the wasm-path-override functions since I didn't end up writing a test for them. The wasm-path-override functions can be added back in later if we want tests for the path override for training.

File: js/web/test/training/e2e/run.js (hunk 4 above)
Comment (carzh) [reply js/web/test/training/e2e/run.js:133]: applied suggestion

File: js/web/test/training/e2e/common.js (hunk 6 above)
Comment (carzh) [reply js/web/test/training/e2e/common.js]: oh good catch; updated with a more thorough test for if two lists are the same

File: js/web/test/training/e2e/run.js (hunk 5 above)
Comment (carzh) [reply js/web/test/training/e2e/run.js]: applied suggestion -- now copies to `{onnxruntime root}/build/js/e2e-training`

**Comment by carzh** [comment]
[Successful pipeline run](https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=1262158&view=results) -- the E2E training tests stage can be seen [here](https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=1262158&view=logs&j=4cf9212c-8936-5e77-cbdb-290c1c5567eb&t=f0659886-20a3-5066-0871-c8b7821377a1)

**Review by jchen351 — APPROVED** [review]
Looks good from the ES side.

**Review by askhade — APPROVED** [review]
LGTM

**Review by snnn — APPROVED** [review]
